pkg/grpc/databroker/leaser.go

```diff
@@ -0,0 +1,130 @@
+package databroker
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/pomerium/pomerium/internal/log"
+)
+
+// A LeaserHandler is a handler for the locker.
+type LeaserHandler interface {
+	GetDataBrokerServiceClient() DataBrokerServiceClient
+	RunLeased(ctx context.Context) error
+}
+
+// A Leaser attempts to acquire a lease and if successful runs the handler. If the lease
+// is released the context used for the handler will be canceled and a new lease
+// acquisition will be attempted.
+type Leaser struct {
+	handler   LeaserHandler
+	leaseName string
+	ttl       time.Duration
+}
+
+// NewLocker creates a new Leaser.
+func NewLocker(leaseName string, ttl time.Duration, handler LeaserHandler) *Leaser {
+	return &Leaser{
+		leaseName: leaseName,
+		ttl:       ttl,
+		handler:   handler,
+	}
+}
+
+// Run acquires the lease and runs the handler. This continues until either:
+//
+// 1. ctx is canceled
+// 2. a non-cancel error is returned from handler
+//
+func (locker *Leaser) Run(ctx context.Context) error {
+	retryTicker := time.NewTicker(locker.ttl / 2)
+	defer retryTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-retryTicker.C:
+		}
+
+		res, err := locker.handler.GetDataBrokerServiceClient().AcquireLease(ctx, &AcquireLeaseRequest{
```

Comment:
I think for the databroker calls we actually want to retry with backoff. We still need to kill the `RunLeased` ctx immediately, but getting disconnected from the databroker seems like something we should handle here. (as opposed to crashing out and relying on the process manager to restart)